### PR TITLE
Feature management interface

### PR DESCRIPTION
**Pull request by plskeggs**
Make changes to support shell changes to gateway BLE connections, and to make other shell commands work better, such as the 'kernel threads' command.  Also fix a few bugs, and expand cJSON.* to offer less memory-hungry *CS versions of various functions.

Reviewers: review by commit, and ignore any commits between:
  d4576af "replace CONFIG_APR_GATEWAY with more general CONFIG_NRF_CLOUD_GATEWAY"
  and
  c764f8e "west: doc: reference correct zephyr / Point to fork and branch for modified zephyr 2.3.0."
those are in an earlier PR (https://github.com/nRFCloud/sdk-nrf/pull/4) which has not been merged.

### CODE REVIEW

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_codec.c [hunk 1]
```diff
@@ -414,6 +503,103 @@ int nrf_cloud_encode_config_response(struct nrf_cloud_data const *const input,
...
+		LOG_ERR("nrf_cloud_encode_gateway_desired_list() failed %d",
+			err);
+		return err;
+	}
+
+	err = nct_cc_send(&msg);
+	if (err) {
+		LOG_ERR("nct_cc_send failed %d", err);
+		nrf_cloud_free((void *)msg.data.ptr);
+		return err;
```
Comment (jayteemo) [inline subsys/net/lib/nrf_cloud/src/nrf_cloud_codec.c:593]: you can remove the two lines after `LOG_ERR`

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_transport.c [hunk 2]
```diff
@@ -354,14 +356,31 @@ int nct_gw_connect(void)
...
+
 void set_gw_rx_topic(char* topic_prefix)
 {
+	char *end_of_stage = strchr(topic_prefix, '/');
+	int len;
+
+	if (end_of_stage) {
+		len = end_of_stage - topic_prefix;
+		if (len >= sizeof(stage)) {
+			len = sizeof(stage) - 1;
```
Comment (jayteemo) [inline subsys/net/lib/nrf_cloud/src/nrf_cloud_transport.c:372]: should this be considered an error?

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_transport.c [hunk 3]
```diff
@@ -479,7 +498,7 @@ static void gw_client_id_get(int at_socket_fd, char *id, size_t id_len)
 			snprintf(id, NRF_CLOUD_CLIENT_ID_LEN + 1, "%s",
 				 "no-psk-ids");
 		}
-		printk("Gateway ID:%s\n", gateway_id);
+		/* printk("Gateway ID:%s\n", gateway_id); */
```
Comment (jayteemo) [inline subsys/net/lib/nrf_cloud/src/nrf_cloud_transport.c:501]: need to keep this?

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_transport.c [hunk 4]
```diff
@@ -949,6 +968,8 @@ int nct_mqtt_connect(void)
 static int publish_get_payload(struct mqtt_client *client, size_t length)
 {
 	if (length > sizeof(nct.payload_buf)) {
+		LOG_WRN("length specified:%zd larger than payload_buf:%zd",
```
Comment (jayteemo) [inline subsys/net/lib/nrf_cloud/src/nrf_cloud_transport.c:971]: `LOG_ERR` ?

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_codec.c [hunk 5]
```diff
@@ -223,8 +224,90 @@ int nrf_cloud_decode_gateway_state(const char *input_ptr,
...
+			ble_address = cJSON_GetObjectItem(address_obj,
+							  "address");
+		}
+
+		if (ble_address != NULL) {
+			bool found = false;
+
+			addr = ble_address->valuestring;
+			for (int j = 0; j < num_cons; j++) {
+				if ((strcmp(addr, cons[j].addr) == 0) &&
```
[the remark this answers is not in the document]
Comment (plskeggs) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_codec.c:250]: Yes, it's considered a new device -- if the address matches but active is false, this is a device which the user must have deleted during the current run of the firmware, and is now adding it back.

Apricity designed the array of desired connections maintained in the firmware (as well as another array of connected devices) as an unordered pool identical structures, where the active flag indicates whether it is currently in use or free.  It's inefficient.

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_codec.c [hunk 6]
```diff
@@ -223,8 +224,90 @@ int nrf_cloud_decode_gateway_state(const char *input_ptr,
...
+	for (int i = 0; i < num_cons; i++) {
+		bool found = false;
+
+		if (!cons[i].active) {
+			continue;
+		}
+
+		for (int j = 0; j < cJSON_GetArraySize(desired_connections_obj);
+		     j++) {
+			item = cJSON_GetArrayItem(desired_connections_obj, j);
```
[the remark this answers is not in the document]
Comment (plskeggs) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_codec.c:276]: good idea

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_transport.c (hunk 2 above)
Comment (plskeggs) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_transport.c:372]: It only affects the accuracy of the `stage` string which is currently only used in the gateway to display to the user, but yeah, I could log an error at least.

**Review by jayteemo — APPROVED** [review]

**Comment by plskeggs** [comment]
NOTE: this PR will not be merged to master right now. Instead, there is a gw-master branch on nrfcloud that will hold merged PRs on top of the related NCS nrf commit. Eventually, the sum of changes to NCS nrf will be separated as needed into logical PRs and then issued against the NCS nrf upstream. This way, while gateway development is going on along side NCS, each new gateway PR does not need to keep stacking on top of others, which makes it harder for reviewers.

PRs after #5 will need to be against gw-master, after it has been caught up to be rebased on top of the correct NCS master.

**Pull request closed without merging**